Implement gateway application command

SQKo opened the pull request:
This is a unified implementation of DiscordPHP with [DiscordPHP-Slash](https://github.com/discord-php/DiscordPHP-Slash) with **Discord v9 API gateway** method, it is merely a copy-paste from the latter library.

The reason I proposed this is due to the future changes from Discord with Application Commands, and your plan in favor of dropping DiscordPHP-Slash to be easily maintained, as well easier for bot coders to integrate. (No hate for the existing discordphp-slash)

~~Since it is a little confusing to decide, I target this for the develop branch instead of master until approved (also to avoid breaking changes with the main branch)~~.

### Main changes
1. No more **RegisterClient**, commands can be now registered through DiscordPHP **application** (as long the bot is authorized with `application.commands` scope) using factory methods.
2. Global commands are now modeled as GlobalCommandRepository `$discord->application->commands`
3. Guild commands are now modeled as GuildCommandRepository `$guild->commands` although I was quite unsure if this pattern is acceptable, it _looked like_ the guild commands repository are not meant for just the bot's commands, i.e. the `application_id` in Guild are meant for the application id of bots created guild (which is rare anyway), I had to workaround the `getRepositoryAttributes()` to override `application_id` with the Bot's application id as it seems another repository does not bind to `:application_id`. However this is the easiest and working approach I could done.
4. Commands are not immediately saved into application or guild repository, they can be freshen/fetch manually when necessary by the coders. It will be automatically cached upon interaction from commands.
4. `registerCommand` is now method of Discord and no longer need to be linked as previously in d.php-slash
5. Filled the part class **Command** (it is empty class in DiscordPHP-Slash), new class **Option** (a copy of Choices with modified structure for [Application Command Option](https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-structure), unsure If I'm doing it right)
6. Copied `handleInteraction` and `handleApplicationCommand` codes into `Event::INTERACTION_CREATE` with some modifications (added cache)
7. New event handlers Application Command Create, Application Command Update, Application Command Delete (seems to work only on Global Commands, untested)

### Files/classes migrated from DiscordPHP-Slash to DiscordPHP
- `src/Discord/Enums/ApplicationCommandOptionType.php` -> `src/Discord/Enums/ApplicationCommandOptionType.php` (New: Mentionable and Number type)
- `src/Discord/Parts/Choices.php` -> `src/Discord/Parts/Interactions/Command/Choice.php`
- `src/Discord/Parts/RegisteredCommand.php` -> `src/Discord/Helpers/RegisteredCommand.php`
- `src/Discord/Parts/Interaction.php` -> `src/Discord/Parts/Interactions/Interaction.php` (~~Merged, with old d.php-slash methods deprecated~~ as the function points to the existing interaction codes, by any mean people get used to the same function and easy to merge for both Message Component and Application Commands)

### Notes
- Registering command for type User and Message apparently just works like Slash commands
- Must use the updated DiscordPHP-Http [endpoint for the corrected application URL](https://github.com/discord-php/DiscordPHP-Http/pull/1)
- Due to the nature of Slash sub commands, it is not possible to implement it with **DiscordClientCommand** (e.g. `/slash sub` and `/slash sub2` cant have `/slash` without sub command)
- I've tested registering commands from various code from the DiscordPHP slash example, codes posted in DiscordPHP's discord server, also the examples given in the [Discord API docs](https://discord.com/developers/docs/interactions/application-commands) with simple acknowledgement
- Still need testing in various forms of interaction reply (i.e. Message builder, embeds)

### To be implemented later
- [x] Command permission overwrites
- [x] Slash command builder, commands currently can be registered with factory methods using traditional array


Sorry for my bad English.. Contact me on DiscordPHP server Robo_N1X#1821 for quick chat.

SQKo commented:
### Codes used for manual test

Put this next to your discord run code, and adjust comment/uncomment as you wish to try

```php
function testCreateGlobalCommand($application)
{
    $command = $application->commands->create([
        'name' => 'ping',
        'description' => 'pong'
    ]);
    return $application->commands->save($command);
}

function testCreateGuildCommand($guild)
{
    $command = $guild->commands->create([
        'name' => 'my_cool_command',
        'description' => 'a cool command'
    ]);
    return $guild->commands->save($command);
}

function testFreshenGlobalCommands($application)
{
    return $application->commands->freshen();
}

function testFreshenGuildCommands($guild)
{
    return $guild->commands->freshen();
}

function testUpdateGuildCommand($guild)
{
    $command = $guild->commands->get('name', 'my_cool_command');
    $command->name = 'newcommandname';
    return $guild->commands->save($command);
}

function testDeleteGuildCommand($guild)
{
    $command = $guild->commands->get('name', 'newcommandname');
    return $guild->commands->delete($command);
}

function testCreateGlobalCommand2($application)
{
    $command = $application->commands->create([
        "name" => "blep",
        "type" => 1,
        "description" => "Send a random adorable animal photo",
        "options" => [
            [
                "name" => "animal",
                "description" => "The type of animal",
                "type" => 3,
                "required" => True,
                "choices" => [
                    [
                        "name" => "Dog",
                        "value" => "animal_dog"
                    ],
                    [
                        "name" => "Cat",
                        "value" => "animal_cat"
                    ],
                    [
                        "name" => "Penguin",
                        "value" => "animal_penguin"
                    ]
                ]
            ],
            [
                "name" => "only_smol",
                "description" => "Whether to show only baby animals",
                "type" => 5,
                "required" => False
            ]
        ]
    ]);
    return $application->commands->save($command);
}

function testCreateGuildUserCommand($guild)
{
    $command = $guild->commands->create([
        'name' => 'High Five',
        'type' => Command::USER
    ]);
    return $guild->commands->save($command);
}

function testCreateSubCommands($apporguild)
{
    $command = $apporguild->commands->create([
        "name" => "permissions",
        "description" => "Get or edit permissions for a user or a role",
        "options" => [
            [
                "name" => "user",
                "description" => "Get or edit permissions for a user",
                "type" => 2, // 2 is type SUB_COMMAND_GROUP
                "options" => [
                    [
                        "name" => "get",
                        "description" => "Get permissions for a user",
                        "type" => 1, // 1 is type SUB_COMMAND
                        "options" => [
                            [
                                "name" => "user",
                                "description" => "The user to get",
                                "type" => 6, // 6 is type USER
                                "required" => true
                            ],
                            [
                                "name" => "channel",
                                "description" => "The channel permissions to get. If omitted, the guild permissions will be returned",
                                "type" => 7, // 7 is type CHANNEL
                                "required" => false
                            ]
                        ]
                    ],
                    [
                        "name" => "edit",
                        "description" => "Edit permissions for a user",
                        "type" => 1,
                        "options" => [
                            [
                                "name" => "user",
                                "description" => "The user to edit",
                                "type" => 6,
                                "required" => true
                            ],
                            [
                                "name" => "channel",
                                "description" => "The channel permissions to edit. If omitted, the guild permissions will be edited",
                                "type" => 7,
                                "required" => false
                            ]
                        ]
                    ]
                ]
            ],
            [
                "name" => "role",
                "description" => "Get or edit permissions for a role",
                "type" => 2,
                "options" => [
                    [
                        "name" => "get",
                        "description" => "Get permissions for a role",
                        "type" => 1,
                        "options" => [
                            [
                                "name" => "role",
                                "description" => "The role to get",
                                "type" => 8, // 8 is type ROLE
                                "required" => true
                            ],
                            [
                                "name" => "channel",
                                "description" => "The channel permissions to get. If omitted, the guild permissions will be returned",
                                "type" => 7,
                                "required" => false
                            ]
                        ]
                    ],
                    [
                        "name" => "edit",
                        "description" => "Edit permissions for a role",
                        "type" => 1,
                        "options" => [
                            [
                                "name" => "role",
                                "description" => "The role to edit",
                                "type" => 8,
                                "required" => true
                            ],
                            [
                                "name" => "channel",
                                "description" => "The channel permissions to edit. If omitted, the guild permissions will be edited",
                                "type" => 7,
                                "required" => false
                            ]
                        ]
                    ]
                ]
            ]
        ]
    ]);
    return $apporguild->commands->save($command);
}

function testCreateGuildMessageCommand($guild)
{
    $command = $guild->commands->create([
        'name' => 'Bookmark',
        'type' => Command::MESSAGE
    ]);
    return $guild->commands->save($command);
}

$discord->on(Event::MESSAGE_CREATE, function (Message $message, Discord $discord) {
    if (!empty($message->author->bot)) return;
    if (!empty($message->author->user->bot)) return;

    if ($message->content == 'com') {
        //testCreateGlobalCommand($discord->application);
        //testCreateGuildCommand($message->guild);

        
        //testCreateGlobalCommand2($discord->application);
        //testCreateGuildUserCommand($message->guild);
        //testCreateSubCommands($message->guild);
        //testCreateGuildMessageCommand($message->guild);

        //testFreshenGlobalCommands($discord->application);
        /*
        testFreshenGuildCommands($message->guild)->done(function (GuildCommandRepository $commands) use ($message) {
            //testUpdateGuildCommand($message->guild);
            testDeleteGuildCommand($message->guild);
        });
        */
        
    }
});

// register global command `/ping` with Discord
$discord->registerCommand('ping', function (Interaction $interaction, Choices $choices) { 
    $interaction->replyWithSource('Pong!');
});

// register guild command `/palace-test` with Discord
$discord->registerCommand('command_name', function (Interaction $interaction, Choices $choices) {
    $guild = $interaction->guild;
    $channel = $interaction->channel;
    $member = $interaction->member;

    //$interaction->acknowledge(); // "THIS INTERACTION FAILED"
    //$interaction->acknowledge(true); // "Palace Bot Is Thinking"

    //$interaction->reply('Hello, world!'); // "Discord\Http\Exceptions\RequestFailedException: Bad Request"
    $interaction->replyWithSource('Hello, world!'); // WORKS AS EXPECTED! replies to the message and shows the source message
});

// register a command `/hello`
$discord->registerCommand('hello', function (Interaction $interaction, Choices $choices) {
    // do some cool stuff here
    // good idea to var_dump interaction and choices to see what they contain

    // once finished, you MUST either acknowledge or reply to a message
    //$interaction->acknowledge(); // acknowledges the message, doesn't show source message
    //$interaction->acknowledge(true); // acknowledges the message and shows the source message

    // to reply to the message
    //$interaction->reply('Hello, world!'); // replies to the message, doesn't show source message
    $interaction->replyWithSource('Hello, world!'); // replies to the message and shows the source message

    // the `reply` methods take 4 parameters: content, tts, embed and allowed_mentions
    // all but content are optional.
    // read the discord developer documentation to see what to pass to these options:
    // https://discord.com/developers/docs/resources/channel#create-message
});

$discord->registerCommand('my_cool_command', function (Interaction $interaction, Choices $choices) use ($discord) {
    // there are a couple fields in $interaction that will return DiscordPHP parts:
    $interaction->guild;
    $interaction->channel;
    $interaction->member;

    // if you don't link DiscordPHP, it will simply return raw arrays

    //$discord->guilds->get('id', 'coolguild')->members->ban(); // do something ?
    $interaction->acknowledge();
});

$discord->registerCommand('High Five', function (Interaction $interaction, Choices $choices) use ($discord) {
    //var_dump($interaction->data);
    $interaction->acknowledge();
});

$discord->registerCommand('Bookmark', function (Interaction $interaction, Choices $choices) use ($discord) {
    //var_dump($interaction->data);
    $interaction->acknowledge();
});
```

davidcole1340 commented on src/Discord/Parts/Interactions/Interaction.php:
```diff
@@ -341,4 +366,150 @@ protected function respond(array $payload, ?Multipart $multipart = null): Extend
+     * @see deleteOriginalResponse()
+     *
+     * @return ExtendedPromiseInterface
+     */
+    public function deleteInitialResponse(): ExtendedPromiseInterface
+    {
+        return $this->deleteOriginalResponse();
+    }
+
+    /**
+     * Updates a follow up message.
+     * Backported for DiscordPHP-Slash
+     *
+     * @param string               $message_id
+     * @param string|null          $content
+     * @param array[]|Embed[]|null $embeds
+     * @param array|null           $allowed_mentions
+     *
+     * @return ExtendedPromiseInterface
+     */
+    public function updateFollowUpMessage(string $message_id, string $content = null, array $embeds = null, array $allowed_mentions = null)
+    {
+        if ($this->type != Interaction::TYPE_APPLICATION_COMMAND) {
+            $this->discord->logger->warning('You can only update follow up messages that occur due to a application command interaction.');
+        }
+
+        if (! $this->responded) {
+            throw new RuntimeException('Cannot create a follow-up message as the interaction has not been responded to.');
+        }
+
+        return (function () use ($message_id, $content, $embeds, $allowed_mentions): ExtendedPromiseInterface {
+            return $this->http->patch(Endpoint::bind(Endpoint::INTERACTION_FOLLOW_UP, $this->application_id, $this->token, $message_id), [
+                'content' => $content,
+                'embeds' => $embeds,
+                'allowed_mentions' => $allowed_mentions,
+            ]);
+        })()->then(function ($response) {
+            return $this->factory->create(Message::class, $response, true);
+        });
+    }
```
I'm not sure if we should backport these methods from the slash command client, it's already going to be a breaking change for users upgrading so perhaps it's just easier to forward them to the new APIs?

SQKo replied:
I'm not sure either, it really depends on the current users of Slash command, aside of MessageBuilder, the method names can be confusing (e.g. updateInitialResponse vs updateOriginalResponse) so i choose to backport them and mark them as deprecated which will help them to move to the new API by time.
In my opinion this can be removed in future (e.g. discordphp v8?).

davidcole1340 replied:
I think it's best to get it over and done with now since it's a major conversion from a separate library. Functions that are missing can be ported over, but changed to use the new features such as `MessageBuilder` (if they are missing).

We can put together a conversion guide for users coming from the slash client which will make it easier.

SQKo replied:
Done, but i keep the updateFollowUpMessage as this it is missing from here, however updated the methods to only accept MessageBuilder

davidcole1340 commented on src/Discord/WebSockets/Events/ApplicationCommandCreate.php:
```diff
@@ -0,0 +1,27 @@
+     */
+    public function handle(Deferred &$deferred, $data): void
+    {
+        $deferred->resolve($this->factory->create(Command::class, $data, true));
```
Does this need to be inserted into a repository (`GlobalCommandRepository` or `GuildCommandRepository`)?

davidcole1340 replied:
Oh I guess it's already inserted if we are creating one?

SQKo replied:
It's already inserted if we are creating one, however this one was for the APPLICATION COMMAND gateway events that seems not mean for bot (i.e. I thought the bot can receive this event if the command was created else where), I'm removing it.

davidcole1340 commented on src/Discord/WebSockets/Event.php:
```diff
@@ -38,6 +38,11 @@ abstract class Event
     public const GUILD_MEMBERS_CHUNK = 'GUILD_MEMBERS_CHUNK';
     public const INTERACTION_CREATE = 'INTERACTION_CREATE';
 
+    // Application Command
+    public const APPLICATION_COMMAND_CREATE = 'APPLICATION_COMMAND_CREATE';
+    public const APPLICATION_COMMAND_UPDATE = 'APPLICATION_COMMAND_UPDATE';
+    public const APPLICATION_COMMAND_DELETE = 'APPLICATION_COMMAND_DELETE';
```
I can't find these in the Discord docs, are they documented somewhere?

davidcole1340 replied:
So it turns out they aren't meant to be sent to bots: https://github.com/discord/discord-api-docs/issues/3690#issuecomment-902915187

This shouldn't matter since the client should know when a command is modified.

SQKo replied:
I see, I thought it was not working.. I will remove this

davidcole1340 commented on src/Discord/WebSockets/Events/InteractionCreate.php:
```diff
@@ -22,7 +24,45 @@ class InteractionCreate extends Event
+            $checkCommand = function ($command) use ($interaction, &$checkCommand) {
+                // Tries to cache the command
+                // possibly the laziest thing ive ever done - stdClass -> array
+                $cmd = json_decode(json_encode($command), true);
```
If I remember correctly this was only done because the slash client didn't use the same `Part` as this library, and encoded with arrays instead of stdclasses. Might be able to get around this now?

SQKo replied:
This is not necessary now, I just realized that caching command from Interaction Data wouldn't be good for commands with sub commands (options), due to interaction data only sending selected options and not all of available options in the command. Commands can be still manually fetched anyway, I'm Removing this.

davidcole1340 reviewed (commented):
Looks good so far. Let me know when you need another review. `develop` is now merged in so this can target `master` now.

SQKo commented:
Can you review the change to Choice.php and Option.php? I converted it to Part and used Collection for the Command.php attributes.

Also moved the RegisteredCommand out of Part folder as this is used internally by registerCommand

SQKo commented:
Ok after some thoughts, researches and observation on other lib, i find that a setListener model would not suit for the case. People will just parse them right from Interaction Create data just like parsing Message Create event.

I've also removed the Choice from registeredCommand callback, as it is being ambiguous with options & sub commands. Another reason is that not all type of commands (e.g. user commands, message commands) have the choice.

So it is ready to use now. I will work on overwrites.

SQKo commented:
My only doubt with overwrite is the class and file naming. Not sure whether that I should name it "GuildApplicationCommandOverwrite" sounds too long so right now I just named it "Overwrite" inside the Guild & Command folders
